### PR TITLE
feat: add network in use check to `compose down`

### LINKED ISSUE #1425: `nerdctl compose down` with orphan container should keep the same behavior with docker
### What is the problem you're trying to solve

When we execute `nerdctl compose down`. nerdctl does not check is three any other containers still in the network and just remove it.

### Describe the solution you'd like

While `docker compose down` will return error and not remove the network. We should keep the same behavior with it.

### Additional context

The orphan container [scenario](https://github.com/containerd/nerdctl/issues/1344#issuecomment-1241475088)

### PR DESCRIPTION
**Pull request by Retrospection**
Signed-off-by: YuRuosheng <371373446@qq.com>

Close #1425

### CODE REVIEW

**Review by Zheaoli — APPROVED** [review]
LGTM

**Comment by djdongjin** [comment]
Can you add an integration test for this scenario?

**Comment by Retrospection** [comment]
> Can you add an integration test for this scenario?

I may try~

File: cmd/nerdctl/compose_down_linux_test.go [hunk 1]
```diff
@@ -24,6 +24,44 @@ import (
...
+	compFull := testutil.NewComposeDir(t, dockerComposeYAMLFull)
+	defer compFull.CleanUp()
+
+	projectName := fmt.Sprintf("nerdctl-compose-test-%d", time.Now().Unix())
+	t.Logf("projectName=%q", projectName)
+
+	// orphanContainer := fmt.Sprintf("%s_orphan_1", projectName)
+
+	base.ComposeCmd("-p", projectName, "-f", compFull.YAMLFullPath(), "up", "-d").AssertOK()
+	// defer base.ComposeCmd("-p", projectName, "-f", compFull.YAMLFullPath(), "down", "-v").Run()
```
Comment (fahedouch) [inline cmd/nerdctl/compose_down_linux_test.go:59]: should be deleted ?

File: pkg/cmd/compose/compose.go [hunk 2]
```diff
@@ -58,6 +58,21 @@ func New(client *containerd.Client, globalOptions types.GlobalCommandOptions, op
 		return false, nil
 	}
 
+	options.NetworkInUse = func(ctx context.Context, netName string) (bool, error) {
+		usedMap, err := netutil.UsedNetworks(ctx, client)
+		if err != nil {
+			return false, err
+		}
+		for _, v := range usedMap {
+			for _, net := range v {
```
Comment (fahedouch) [inline pkg/cmd/compose/compose.go:67]: `v` here is a container slice not a network slice, right ?

File: pkg/composer/down.go [hunk 3]
```diff
@@ -97,6 +97,15 @@ func (c *Composer) downNetwork(ctx context.Context, shortName string) error {
 	if err != nil {
 		return err
 	} else if netExists {
+		netUsed, err := c.NetworkInUse(ctx, fullName)
+		if err != nil {
+			return err
+		}
+		if netUsed {
+			logrus.Errorf("network %s is in use", fullName)
+			return nil
```
Comment (fahedouch) [inline pkg/composer/down.go:106]: why not `return fmt.Errorf("network %s is in use", fullName)` ?

**Review by fahedouch — CHANGES_REQUESTED** [review]

File: cmd/nerdctl/compose_down_linux_test.go [hunk 4]
```diff
@@ -24,6 +24,44 @@ import (
...
+
+	base.ComposeCmd("-p", projectName, "-f", compOrphan.YAMLFullPath(), "down", "-v").AssertCombinedOutContains("is in use")
+	// base.ComposeCmd("-p", projectName, "-f", compFull.YAMLFullPath(), "ps").AssertOutNotContains(orphanContainer)
```
Comment (fahedouch) [inline cmd/nerdctl/compose_down_linux_test.go:62]: would you please explain your intention behind this integration test ? Thanks

**Comment by AkihiroSuda** [comment]
> Signed-off-by: Retros

Please use a real name, as explained in https://github.com/cncf/foundation/issues/383#issuecomment-1178254458

File: cmd/nerdctl/compose_down_linux_test.go (hunk 1 above)
Comment (Retrospection) [reply cmd/nerdctl/compose_down_linux_test.go:59]: yep，I pushed this commit just want to try to see if I can pass the CI test~ 

Will refactor this

File: cmd/nerdctl/compose_down_linux_test.go (hunk 4 above)
Comment (Retrospection) [reply cmd/nerdctl/compose_down_linux_test.go:62]: I make this  integration test to ensure when run `compose down` command while the network is in use, the stderr will output error message "network xxxxx is in use". This behavior proves that the network in use inspection is working.

File: pkg/cmd/compose/compose.go (hunk 2 above)
Comment (Zheaoli) [reply pkg/cmd/compose/compose.go:67]: Yes，I think we should use the key as the network name

Comment (Retrospection) [reply pkg/cmd/compose/compose.go:67]: > `v` here is a container slice not a network slice, right ?

nope, `useMap` is a map which key is namespace and value is network names.

**Comment by suyanhanx** [comment]
Please also update your DCO sign in the PR description.

File: cmd/nerdctl/compose_down_linux_test.go [hunk 5]
```diff
@@ -24,6 +24,43 @@ import (
...
+	compOrphan := testutil.NewComposeDir(t, dockerComposeYAMLOrphan)
+	defer compOrphan.CleanUp()
+	compFull := testutil.NewComposeDir(t, dockerComposeYAMLFull)
+	defer compFull.CleanUp()
+
+	projectName := fmt.Sprintf("nerdctl-compose-test-%d", time.Now().Unix())
+	t.Logf("projectName=%q", projectName)
+
+	base.ComposeCmd("-p", projectName, "-f", compFull.YAMLFullPath(), "up", "-d").AssertOK()
+	base.ComposeCmd("-p", projectName, "-f", compOrphan.YAMLFullPath(), "down", "-v").AssertCombinedOutContains("is in use")
```
Comment (suyanhanx) [inline cmd/nerdctl/compose_down_linux_test.go:61]: Need to clean up the orphan container.

**Pull request closed without merging**